File: 1301-1400/1340-Jump Game V/1340-Jump Game V.py
from functools import lru_cache
class Solution:
    def maxJumps(self, arr: List[int], d: int) -> int:        
        @lru_cache(None)
        def dfs(curr):
            currJump = 1
            for next in range(curr - 1, max(0, curr - d) - 1, -1):
                if arr[curr] > arr[next]:
                    currJump = max(currJump, 1 + dfs(next))
                else:
                    break
            for next in range(curr + 1, min(len(arr) - 1, curr + d) + 1):
                if arr[curr] > arr[next]:
                    currJump = max(currJump, 1 + dfs(next))
                else:
                    break
            return currJump
        maxJ = 0
        for i in range(len(arr)):
            maxJ = max(maxJ, dfs(i))
        return maxJ

File: 1301-1400/1340-Jump Game V/1340-Jump Game V.cpp
class Solution {
public:
    int maxJumps(vector<int>& arr, int d) {
        int maxJ = 0;
        vector<int> table(arr.size());
        for (int i = 0; i < arr.size(); ++i) {
            maxJ = max(maxJ, dfs(arr, i, d, table));
        }
        return maxJ;
    }
    
private:
    int dfs(vector<int>& arr, int curr, int d, vector<int>& table) {
        if (table[curr] == 0) {
            int currJump = 1;
            for (int next = curr - 1; next >= max(0, curr - d); --next) {
                if (arr[curr] > arr[next]) {
                    currJump = max(currJump, 1 + dfs(arr, next, d, table));
                }
                else {
                    break;
                }
            }
            
            for (int next = curr + 1; next <= min(int(arr.size()) - 1, curr + d); ++next) {
                if (arr[curr] > arr[next]) {
                    currJump = max(currJump, 1 + dfs(arr, next, d, table));
                }
                else {
                    break;
                }
            }
            table[curr] = currJump;
        }
        return table[curr];
    }
};
